dcos/mesos.py

```diff
@@ -912,6 +976,410 @@ def __str__(self):
+        self.args = args
+
+        # Create a client and grab a reference to the DC/OS master.
+        client = DCOSClient()
+        master = get_master(client)
+
+        # Get the URL to the agent running the task.
+        task_obj = master.task(task_id)
+        if client._mesos_master_url:
+            self.agent_url = client.slave_url(
+                "", task_obj.slave().http_url(), "api/v1")
+        else:
+            self.agent_url = client.slave_url(
+                task_obj.slave()['id'], "", "api/v1")
+
+        # Grab a reference to the container ID for the task.
+        self.parent_id = master.get_container_id(task_id)
+
+        # Generate a new UUID for the nested container
+        # used to run commands passed to `task exec`.
+        self.container_id = str(uuid.uuid4())
+
+        # Set up a recordio encoder and decoder
+        # for any incoming and outgoing messages.
+        self.encoder = recordio.Encoder(
+            lambda s: bytes(json.dumps(s, ensure_ascii=False), "UTF-8"))
+        self.decoder = recordio.Decoder(
+            lambda s: json.loads(s.decode("UTF-8")))
+
+        # Set up queues to send messages between threads used for
+        # reading/writing to STDIN/STDOUT/STDERR and threads
+        # sending/receiving data over the network.
+        self.input_queue = Queue()
+        self.output_queue = Queue()
+
+        # Set up an event to block attaching
+        # input until attaching output is complete.
+        self.attach_input_event = threading.Event()
+        self.attach_input_event.clear()
+
+        # Set up an event to block the main thread
+        # from exiting until signaled to do so.
+        self.exit_event = threading.Event()
+        self.exit_event.clear()
+
+        # Use a class variable to store exceptions thrown on
+        # other threads and raise them on the main thread before
+        # exiting.
+        self.exception = None
+
+    def run(self):
+        """Run the helper threads in this class which enable streaming
+        of STDIN/STDOUT/STDERR between the CLI and the Mesos Agent API.
+
+        If a tty is requested, we take over the current terminal and
+        put it into raw mode. We make sure to reset the terminal back
+        to its original settings before exiting.
+        """
+
+        # Without a TTY.
+        if not self.tty:
+            try:
+                self._start_threads()
+                self.exit_event.wait()
+            except Exception as e:
+                self.exception = e
+
+            if self.exception:
+                raise self.exception
+            return
+
+        # With a TTY.
+        if util.is_windows_platform():
+            raise DCOSException(
+                "Running with the '--tty' flag is not supported on windows.")
+
+        if not sys.stdin.isatty():
+            raise DCOSException(
+                "Must be running in a tty to pass the '--tty flag'.")
+
+        import termios
+        import tty
```

Comment:
I can't move them to the top of the file because they don't exist on windows (which is why they are below the check for windows. I could do the conditional check for windows at the top as well and then do the import there. Is that preferred?